---
title: "Differential Expression"
author: "Peter Searle"
date: "3/29/2022"
output: html_document
---

# Setup

## Working Directory

The following codes sets the working directory for all subsequent code chunks

```{r setup, include=FALSE, echo=FALSE}

require("knitr")
opts_knit$set(root.dir = "/home/searleps/github/sucker/differential_expression/")

```

## Installation of Miniconda3

```{bash eval=FALSE}
#!/bin/bash
wget https://repo.anaconda.com/miniconda/Miniconda3-py39_4.9.2-Linux-x86_64.sh
bash Miniconda3-py39_4.9.2-Linux-x86_64.sh 
source ~/miniconda3/bin/activate 
```

# Predict Gene Names

Predict gene names for transcriptomes using EggNOG Mapper.

```{bash}
#!/bin/bash

# Installation

#conda create -n eggnog-mapper
#conda activate eggnog-mapper
#conda install -c bioconda -c conda-forge eggnog-mapper
#mkdir ~/miniconda3/envs/eggnog-mapper/lib/python3.11/site-packages/data
#download_eggnog_data.py

cd ~/github/sucker/differential_expression/data/predicted_genes

for lineage in june utah straw;
do
  mkdir $lineage
  cd $lineage
  mkdir logs
  emapper.py --cpu 20 -o out --data_dir ~/miniconda3/lib/python3.11/site-packages/data \
  --temp_dir ~/github/sucker/differential_expression/data/predicted_genes/${lineage}/ --override \
  -m diamond --dmnd_ignore_warnings -i ../transcriptomes/unfiltered/${lineage}_transcripts.fasta \
  --evalue 0.001 --score 60 --pident 40 --query_cover 20 --subject_cover 20 --itype genome \
  --genepred search --tax_scope 7898 --target_orthologs all --go_evidence non-electronic \
  --pfam_realign none --report_orthologs --decorate_gff yes --excel \
  > ${lineage}/logs/emapper.out 2> ${lineage}/logs/emapper.err
  cd ..
done

```

# Filter Transcriptomes

## Filtered Transcripts .lst File

Generate list of transcripts that are annotated and share gene names across all lineages. 

```{r}
library(tidyverse)
library(dplyr)

# File Paths

june_path <- "./data/predicted_genes/june/out.emapper.annotations.xlsx"
utah_path <- "./data/predicted_genes/utah/out.emapper.annotations.xlsx"
straw_path <- "./data/predicted_genes/straw/out.emapper.annotations.xlsx"

# Filtered transcripts 

june_fltrd <- readxl::read_excel(june_path, skip = 2) %>% 
  dplyr::rename(query = "query", gene = "Preferred_name", evalue = "evalue", score = "score") %>%
  drop_na() %>% select(query, gene, evalue, score) %>% # remove last 3 rows
  mutate(across('query', str_replace, '_\\d', '')) %>% # rename transcripts
  mutate(evalue = as.numeric(evalue)) %>% filter(gene != "-") %>% #remove genes without names
  group_by(query) %>% slice_min(order_by = evalue) %>% 
  # transcripts can have multiple annotations. Only keep one based off min evalue
  mutate(geneL = tolower(gene))

utah_fltrd <- readxl::read_excel(utah_path, skip = 2) %>% 
  dplyr::rename(query = "query", gene = "Preferred_name", evalue = "evalue", score = "score") %>%
  drop_na() %>% select(query, gene, evalue, score) %>% # remove last 3 rows
  mutate(across('query', str_replace, '_\\d', '')) %>% # rename transcripts
  mutate(evalue = as.numeric(evalue)) %>% filter(gene != "-") %>% #remove genes without names
  group_by(query) %>% slice_min(order_by = evalue) %>% # only keep one annotation per transcript based off minimum evalue
  mutate(geneL = tolower(gene))

straw_fltrd <- readxl::read_excel(straw_path, skip = 2) %>% 
  dplyr::rename(query = "query", gene = "Preferred_name", evalue = "evalue", score = "score") %>%
  drop_na() %>% select(query, gene, evalue, score) %>% # remove last 3 rows
  mutate(across('query', str_replace, '_\\d', '')) %>% # rename transcripts
  mutate(evalue = as.numeric(evalue)) %>% filter(gene != "-") %>% #remove genes without names
  group_by(query) %>% slice_min(order_by = evalue) %>% # only keep one annotation per transcript based off minimum eval
  mutate(geneL = tolower(gene))

# Common genes between 3 lineages

common_genes <- as_vector(intersect(intersect(june_fltrd$geneL, utah_fltrd$geneL), straw_fltrd$geneL))



# Next steps

# In the case where there are two gene names that are the same except for the case differences (for example,mmp2 vs MMP2), how can I change only those instances to lower case? 
#Min eValue or max eValue....this is not very clear to me!

common_genes[duplicated(common_genes)]

common_genes_cnt <- as_tibble(common_genes) %>% mutate(value = toupper(value)) %>% summarise(count = n_distinct(value))

test <- unique(june_fltrd$gene) %>% toupper()

test[duplicated(test)]

june_fltrd_up <- mutate(june_fltrd, gene = tolower(gene)) %>% summarise(count = n_distinct(gene))
june_fltrd$gene[duplicated(june_fltrd$gene)]


# Write tables


write.table(common_genes, "./data/predicted_genes/common_genes.txt", row.names = FALSE, col.names = FALSE, quote = FALSE)

june_final <- june_eValue %>%
  filter(geneName %in% common_genes)

utah_final <- utah_eValue %>%
  filter(geneName %in% common_genes)

straw_final <- straw_eValue %>%
  filter(geneName %in% common_genes)

path <- "./data/transcriptomes/filtered/transcript_IDs/"

# Statistics
results <- tibble(Category = "All", June = nrow(june_all), Utah = nrow(utah_all), Straw = nrow(straw_all))
results <- add_row(results, Category = "NA Removed", June = nrow(june_na_removed), Utah = nrow(utah_na_removed), Straw = nrow(straw_na_removed))
results <- add_row(results, Category = "eValue (1e-30)", June = nrow(june_eValue), Utah = nrow(utah_eValue), Straw = nrow(straw_eValue))
results <- add_row(results, Category = "Final", June = nrow(june_final), Utah = nrow(utah_final), Straw = nrow(straw_final))
write_tsv(results, paste0(path, "filter_results.tsv"))

write.table(june_final$queryID, paste0(path, "june_transcripts.lst"), row.names = FALSE)
write.table(utah_final$queryID, paste0(path, "utah_transcripts.lst"), row.names = FALSE)
write.table(straw_final$queryID, paste0(path, "straw_transcripts.lst"), row.names = FALSE)
```

Modify text in transcripts.lst files so seqtk will run.

```{bash}
#!/bin/bash
source ~/miniconda3/bin/activate 

cd ~/github/sucker/differential_expression/data/transcriptomes/filtered/transcript_IDs/

# Remove x and "'s from transcripts.lst 

sed -i 's/x//g' *.lst
sed -i 's/\"//g' *.lst

# Remove blank line at start of file

sed -i '/^$/d' *.lst
```

## Filtered Transcripts .fasta File

Use seqtk to filter assembled transcriptomes using parameters described above. The unfiltered transcriptomes are available in this repository under data/transcriptomes/unfiltered/

```{bash}
#!/bin/bash
source ~/miniconda3/bin/activate 

# Installation 

conda create -n seqtk seqtk
conda activate seqtk

# seqtk

cd ~/github/sucker/differential_expression/data/transcriptomes/filtered/transcript_IDs/

seqtk subseq ~/github/sucker/differential_expression/data/transcriptomes/unfiltered/june_transcripts.fasta june_transcripts.lst > ../filtered_transcripts/june_filtered_transcripts.fasta
seqtk subseq ~/github/sucker/differential_expression/data/transcriptomes/unfiltered/utah_transcripts.fasta utah_transcripts.lst > ../filtered_transcripts/utah_filtered_transcripts.fasta
seqtk subseq ~/github/sucker/differential_expression/data/transcriptomes/unfiltered/straw_transcripts.fasta straw_transcripts.lst > ../filtered_transcripts/straw_filtered_transcripts.fasta
```

## Generate first transcript to gene mapping file

This file is used in the filtering step to associate each transcript with its gene annotation

```{r}
library(tidyverse)

june_tx2gene <- select(june_final, queryID, geneName)
colnames(june_tx2gene) <- c('txID', 'geneID')

utah_tx2gene <- select(utah_final, queryID, geneName)
colnames(utah_tx2gene) <- c('txID', 'geneID')

straw_tx2gene <- select(straw_final, queryID, geneName)
colnames(straw_tx2gene) <- c('txID', 'geneID')

path <- "./data/transcriptomes/filtered/tx2gene/"

write_tsv(june_tx2gene, paste0(path, "june_tx2gene.tsv"), col_names = T)
write_tsv(utah_tx2gene, paste0(path, "utah_tx2gene.tsv"), col_names = T)
write_tsv(straw_tx2gene, paste0(path, "straw_tx2gene.tsv"), col_names = T)
```

## Final Transcriptomes .fasta File

Additional filtering is required in order to make the transcriptomes comparable among lineages at the transcript-level (i.e. each gene has the same number of transcripts represented for each lineage and each of these transcripts are similar between lineages).

```{bash}
#!/bin/bash

# Issue with gene named '26-29-p' with seqtk. Renaming it solves the problem. Filtering ends of removing this gene in the end, so it doesn't need to be renamed later.

cd ~/github/sucker/differential_expression/data/transcriptomes/filtered/tx2gene/

sed -i 's/26-29-p/twentysixtwentyninep/g' *.tsv

cd ~/github/sucker/differential_expression/data/predicted_genes/

sed -i 's/26-29-p/twentysixtwentyninep/g' common_genes.txt

```

Use seqtk to pull all transcripts annotated as a gene from all lineages and combine into one fasta file

```{bash}
#!/bin/bash

#conda create -n seqtk 
#conda install seqtk
#conda install seqkit

conda activate seqtk

cd ~/github/sucker/differential_expression/data/transcriptomes/filtered/multi_transcripts/

> logs/seqtk1.log

printf "Extracting transcripts using seqtk for each gene\n" >> logs/seqtk1.log

count=1

cat ../../../predicted_genes/common_genes_test.txt | while read gene;
do
  printf "Processing $gene $count\n" >> logs/seqtk1.log
  let count++
  > ${gene}.unfltr.fa
  for file in ../tx2gene/*.tsv;
  do
    samp=`basename ${file} _tx2gene.tsv`
    grep -w $gene $file > gene.txt
    seqtk subseq ../filtered_transcripts/${samp}_filtered_transcripts.fasta \
    gene.txt | sed -r 's/(>.+)(\s.+)/\1_'"${samp}"'_'"${gene}"'/g' >> ${gene}.unfltr.fa
    done
    rm gene.txt
done

```

Cluster transcripts using cd-hit-est

```{bash}
#!/bin/bash

#conda create -n cd-hit cd-hit

conda activate cd-hit

> logs/cd_hit_est_1.log

printf "Generating clusters with cd-hit-est\n"

count=1

for file in *.unfltr.fa
do
  gene=`basename ${file} .unfltr.fa`
  printf "Processing $gene $count\n" >> logs/cd_hit_est_1.log
  let count++
  cd-hit-est -i $file -o $gene -c .9 -n 10 -d 100 >> logs/cd_hit_est_2.log
  ~/miniconda3/envs/cd-hit/bin/clstr2txt.pl ${gene}.clstr > ${gene}.clstr.txt
  rm ${gene}.clstr
  rm ${gene}
  mv ${gene}.clstr.txt ${gene}.clstr
done
```

Determine which clusters have all three lineages represented and all transcripts in each cluster with clstr identity >= 97%. If there are multiple instances of transcripts in a cluster from the same lineage, only retain one transcript with highest clstr identity per lineage

```{bash}
#!/bin/bash

#conda create -n pandas pandas
conda activate pandas

> logs/pandas.log

printf "Filtering clusters with python\n" >> logs/pandas.log

count=1

for file in *.clstr
do
  gene=`basename ${file} .clstr`
  printf "Processing $gene $count\n" >> logs/pandas.log
  let count++
  ~/github/sucker/differential_expression/code/cluster.py $file ${gene}.clstr.srtd
done

```

Generate filtered fasta files based off parameters described above 

```{bash}
#!/bin/bash
conda activate seqtk

> logs/seqtk2.log

printf "Filtering fasta files using seqtk and renaming fasta file headers\n" >> logs/seqtk2.log

count=1

for file in *.clstr.srtd;
do
  gene=`basename ${file} .clstr.srtd`
  printf "Processing $gene $count\n" >> logs/seqtk2.log
  let count++
  seqtk subseq ${gene}.unfltr.fa ${file} > ${gene}.unfltrd.fa
  seqkit replace -p "^(\S+)" -r '{kv}' -k ${gene}.clstr.srtd ${gene}.unfltrd.fa > ${gene}.fltrd.fa 
  awk '!/^>/ { printf "%s", $0; n = "\n" } /^>/ { print n $0; n = "" } END { printf "%s", n }' ${gene}.fltrd.fa > \
  ${gene}.fltr.fa 
  #rm ${gene}.fltrd.fa
  #rm ${gene}.clstr.srtd
  #rm ${gene}.unfltrd.fa
  #rm ${gene}.unfltr.fa
  #rm ${gene}.clstr
done
```

Create final transcriptomes for all three lineages. 

```{bash}
#!/bin/bash

> logs/transcriptomes.log

printf "Generating final transcriptomes\n" >> logs/transcriptomes.log

for lineage in june utah straw;
do
  > ../final_transcripts/${lineage}.fa
done

count=1

for file in *.fltr.fa
do
  gene=`basename ${file} .fltr.fa`
  printf "Processing $gene $count\n" >> logs/transcriptomes.log
  let count++
  for lineage in june utah straw;
  do
    grep $lineage $file > gene.txt
    grep -w -f gene.txt -A 1 $file >> ../final_transcripts/${lineage}.fa
    done
  rm gene.txt
  #rm ${gene}.fltr.fa
done
```


# Quantify Transcript Abundance with Salmon

```{bash}
#!/bin/bash
source ~/miniconda3/bin/activate 

# Install salmon in its own environment

conda config --add channels conda-forge
conda config --add channels bioconda
conda create -n salmon salmon

# Activate salmon environment

conda activate salmon
```

## Build an Index

Build an index of the transcriptomes before performing transcript-level quantification

```{bash}
#!/bin/bash
source ~/miniconda3/bin/activate
conda activate salmon

cd ~/github/sucker/differential_expression/data/salmon/index/

salmon index -t ~/github/sucker/differential_expression/data/transcriptomes/filtered/final_transcripts/june.fa -i june_index
salmon index -t ~/github/sucker/differential_expression/data/transcriptomes/filtered/final_transcripts/utah.fa -i utah_index
salmon index -t ~/github/sucker/differential_expression/data/transcriptomes/filtered/final_transcripts/straw.fa -i straw_index
```

## Quantify Samples

Salmon then quantifies transcript-level abundance using the index and RNA-seq data. The RNA-seq data is available under BioProject PRJNA818778.

June sucker

```{bash}
#!/bin/bash
source ~/miniconda3/bin/activate
conda activate salmon

for fn in ~/github/sucker/differential_expression/data/rna_seq/june/*_1.fq.gz;
do
  samp=`basename ${fn} _1.fq.gz`
  echo "Processing sample ${samp}"

  dir=`dirname ${fn}`
  echo "${dir}"
  
  salmon quant -i ~/github/sucker/differential_expression/data/salmon/index/june_index -l A \
    -1 $dir/${samp}_1.fq.gz \
    -2 $dir/${samp}_2.fq.gz \
    -p 8 --validateMappings --gcBias \
    -o ~/github/sucker/differential_expression/data/salmon/quant/june/${samp}

done
```

Utah Lake Utah sucker

```{bash}
#!/bin/bash
source ~/miniconda3/bin/activate
conda activate salmon

for fn in ~/github/sucker/differential_expression/data/rna_seq/utah/*_1.fq.gz;
do
  samp=`basename ${fn} _1.fq.gz`
  echo "Processing sample ${samp}"

  dir=`dirname ${fn}`
  echo "${dir}"
  
  salmon quant -i ~/github/sucker/differential_expression/data/salmon/index/utah_index -l A \
    -1 $dir/${samp}_1.fq.gz \
    -2 $dir/${samp}_2.fq.gz \
    -p 8 --validateMappings --gcBias \
    -o ~/github/sucker/differential_expression/data/salmon/quant/utah/${samp}

done
```

Strawberry Reservoir Utah sucker

```{bash}
#!/bin/bash
source ~/miniconda3/bin/activate
conda activate salmon

for fn in ~/github/sucker/differential_expression/data/rna_seq/straw/*_1.fq.gz;
do
  samp=`basename ${fn} _1.fq.gz`
  echo "Processing sample ${samp}"

  dir=`dirname ${fn}`
  echo "${dir}"
  
  salmon quant -i ~/github/sucker/differential_expression/data/salmon/index/straw_index -l A \
    -1 $dir/${samp}_1.fq.gz \
    -2 $dir/${samp}_2.fq.gz \
    -p 8 --validateMappings --gcBias \
    -o ~/github/sucker/differential_expression/data/salmon/quant/straw/${samp}

done
```

# Generate DESeqDataSet

## Generate second transcript to gene mapping file

This file associates transcripts with gene IDs so that we can quantify expression at the gene-level. 

```{bash}
#!/bin/bash

cd ~/github/sucker/differential_expression/data/transcriptomes/filtered/final_transcripts

for file in *.fa;
do
  lineage=`basename ${file} .fa`
  grep ">" ${lineage}.fa | perl -pe 's/>(\w+\/\d+)_(\w+)_(.+)(_)(\d+)/$1\t$2\t$3\t$3$4$5/' \ 
  > ../tx2gene/${lineage}_tx2gene.fltrd.tsv
done

```


```{r}
library(tidyverse)

path <- "./data/transcriptomes/filtered/tx2gene/"

june_tx2gn <- read_delim(paste0(path, "june_tx2gene.fltrd.tsv"), col_names = FALSE)
utah_tx2gn <- read_delim(paste0(path, "utah_tx2gene.fltrd.tsv"), col_names = FALSE)
straw_tx2gn <- read_delim(paste0(path, "straw_tx2gene.fltrd.tsv"), col_names = FALSE)

colnames(june_tx2gn) <- c("txID", "species", "geneID", "txForm")
colnames(utah_tx2gn) <- c("txID", "species", "geneID", "txForm")
colnames(straw_tx2gn) <- c("txID", "species", "geneID", "txForm")

june_tx2gene1 <- select(june_tx2gn, txID, geneID)
utah_tx2gene1 <- select(utah_tx2gn, txID, geneID)
straw_tx2gene1 <- select(straw_tx2gn, txID, geneID)
```

## Transcript abundance files and tximport

Import transcript quantification data using tximport and create gene-level count tables 

```{r}
library(readr)
library(tximport)

# Create path to directory including sample files

june_dir <- "./data/sample_info"
utah_dir <- "./data/sample_info"
straw_dir <- "./data/sample_info"

# Manually create and import table with samples and time columns

june_samples <- read.table(file.path(june_dir, "june_samples.txt"), header = TRUE)
utah_samples <- read.table(file.path(utah_dir, "utah_samples.txt"), header = TRUE)
straw_samples <- read.table(file.path(straw_dir, "straw_samples.txt"), header = TRUE)

# Add rownames to samples table

rownames(june_samples) <- june_samples$Sample
rownames(utah_samples) <- utah_samples$Sample
rownames(straw_samples) <- straw_samples$Sample

# Create path to directory including quant.sf files

june_dir <- "./data/salmon/quant/june"
utah_dir <- "./data/salmon/quant/utah"
straw_dir <- "./data/salmon/quant/straw"

# Create a named vector pointing to the quantification files
 
june_files <- file.path(june_dir, june_samples$Sample, "quant.sf")
june_names <- paste0(june_samples$Sample)
names(june_files) <- june_names
print("June Files")
all(file.exists(june_files))

utah_files <- file.path(utah_dir, utah_samples$Sample, "quant.sf")
utah_names <- paste0(utah_samples$Sample)
names(utah_files) <- utah_names
print("Utah Files")
all(file.exists(utah_files))

straw_files <- file.path(straw_dir, straw_samples$Sample, "quant.sf")
straw_names <- paste0(straw_samples$Sample)
names(straw_files) <- straw_names
print("Straw Files")
all(file.exists(straw_files))

# Import quantification data for DESeq2 using tximport
june_txi <- tximport(june_files, type = "salmon", tx2gene = june_tx2gene)
utah_txi <- tximport(utah_files, type = "salmon", tx2gene = utah_tx2gene)
straw_txi <- tximport(straw_files, type = "salmon", tx2gene = straw_tx2gene)

# Confirm number of rows is the same between different objects
nrow(june_txi$counts) == 10006
nrow(utah_txi$counts) == 10006
nrow(straw_txi$counts) == 10006
```

## Construct a DESeqDataSet

```{r}
library(DESeq2)

# Set Time as factor
june_samples$Time <- factor(june_samples$Time)
utah_samples$Time <- factor(utah_samples$Time)
straw_samples$Time <- factor(straw_samples$Time)

# Generate DESeqDataSet 
june_dds <- DESeqDataSetFromTximport(june_txi,
                              colData = june_samples,
                              design = ~ Time)
utah_dds <- DESeqDataSetFromTximport(utah_txi,
                              colData = utah_samples,
                              design = ~ Time)
straw_dds <- DESeqDataSetFromTximport(straw_txi,
                              colData = straw_samples,
                              design = ~ Time)
```

# Data Transformation and Visualization

## Transformation

Transform count data to aid in visualization 

```{r}
#RLD Transformation
june_rld <- rlog(june_dds, blind=TRUE)
utah_rld <- rlog(utah_dds, blind=TRUE)
straw_rld <- rlog(straw_dds, blind=TRUE)

```

## Visualization with Principal Component Plot

Perform quality assessment of data prior to differential expression analyses. The PCA plot is particularly useful for identifying potential outliers. 

```{r}
library(ggrepel)
library(ggpubr)

june_pca_rld <- plotPCA(june_rld, intgroup="Time") +
  labs(title = "June") +
  theme_bw() +
  theme(plot.title = element_text(hjust = .5, size = 10),
        legend.position = "none", 
        axis.title = element_text(size = 7), 
        axis.text = element_text(size = 5),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()
        )

utah_pca_rld <- plotPCA(utah_rld, intgroup="Time") +
  labs(title = "Utah") +
  theme_bw() +
  theme(plot.title = element_text(hjust = .5, size = 10),
        legend.position = "none", 
        axis.title = element_text(size = 7), 
        axis.text = element_text(size = 5),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()
        )

straw_pca_rld <- plotPCA(straw_rld, intgroup="Time") +
  labs(title = "Strawberry") +
  theme_bw() +
  theme(plot.title = element_text(hjust = .5, size = 10),
        legend.position = "none", 
        axis.title = element_text(size = 7), 
        axis.text = element_text(size = 5),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()
        )

june_pca_rld
utah_pca_rld
straw_pca_rld

PCA_Plot <- ggarrange(june_pca_rld, utah_pca_rld, straw_pca_rld, ncol=1, nrow=3, common.legend = TRUE, legend="bottom")

ggsave("./figures/PCA/PCA.pdf", PCA_Plot, height = 8, width = 4)
```

# Differential Expression Analysis

Run differential expression analysis using DESeq 

```{r}
# Run Differential Expression Analysis

june_dds <- DESeq(june_dds)
utah_dds <- DESeq(utah_dds)
straw_dds <- DESeq(straw_dds)

# Generate DE results by specifying contrast of interest and associated alpha and lfcThreshold values

alpha = .05
lfcThreshold = 0

june_res_12v9 <- results(june_dds, contrast=c("Time","12","9"), alpha = alpha, lfcThreshold = lfcThreshold)
utah_res_12v9 <- results(utah_dds, contrast=c("Time","12","9"), alpha = alpha, lfcThreshold = lfcThreshold)
straw_res_12v9 <- results(straw_dds, contrast=c("Time","12","9"), alpha = alpha, lfcThreshold = lfcThreshold)

june_res_14v9 <- results(june_dds, contrast=c("Time","14","9"), alpha = alpha, lfcThreshold = lfcThreshold)
utah_res_14v9 <- results(utah_dds, contrast=c("Time","14","9"), alpha = alpha, lfcThreshold = lfcThreshold)
straw_res_14v9 <- results(straw_dds, contrast=c("Time","14","9"), alpha = alpha, lfcThreshold = lfcThreshold)

june_res_14v12 <- results(june_dds, contrast=c("Time","14","12"), alpha = alpha, lfcThreshold = lfcThreshold)
utah_res_14v12 <- results(utah_dds, contrast=c("Time","14","12"), alpha = alpha, lfcThreshold = lfcThreshold)
straw_res_14v12 <- results(straw_dds, contrast=c("Time","14","12"), alpha = alpha, lfcThreshold = lfcThreshold)
```

## p-values and adjusted p-values

Summarize basic information with summary function.

```{r}
# Summarize results

summary(june_res_12v9)
summary(utah_res_12v9)
summary(straw_res_12v9)

summary(june_res_14v9)
summary(utah_res_14v9)
summary(straw_res_14v9)

summary(june_res_14v12)
summary(utah_res_14v12)
summary(straw_res_14v12)
```

Check how many samples had adjusted p-values less than alpha, adjusted p-values != NA, and abs(log2fold change) > lfcThreshold?

```{r}
# Adjusted p-values for 12v9 contrast

print("June 12v9")
sum(june_res_12v9$padj < alpha & abs(june_res_12v9$log2FoldChange) > lfcThreshold, na.rm=TRUE)
print("Utah 12v9")
sum(utah_res_12v9$padj < alpha & abs(utah_res_12v9$log2FoldChange) > lfcThreshold, na.rm=TRUE)
print("straw 12v9")
sum(straw_res_12v9$padj < alpha & abs(straw_res_12v9$log2FoldChange) > lfcThreshold, na.rm=TRUE)

# Adjusted p-values for 14v9 contrast

print("June 14v9")
sum(june_res_14v9$padj < alpha & abs(june_res_14v9$log2FoldChange) > lfcThreshold, na.rm=TRUE)
print("Utah 14v9")
sum(utah_res_14v9$padj < alpha & abs(utah_res_14v9$log2FoldChange) > lfcThreshold, na.rm=TRUE)
print("straw 14v9")
sum(straw_res_14v9$padj < alpha & abs(straw_res_14v9$log2FoldChange) > lfcThreshold, na.rm=TRUE)

# Adjusted p-values for 14v12 contrast

print("June 14v12")
sum(june_res_14v12$padj < alpha & abs(june_res_14v12$log2FoldChange) > lfcThreshold, na.rm=TRUE)
print("Utah 14v12")
sum(utah_res_14v12$padj < alpha & abs(utah_res_14v12$log2FoldChange) > lfcThreshold, na.rm=TRUE)
print("straw 14v12")
sum(straw_res_14v12$padj < alpha & abs(straw_res_14v12$log2FoldChange) > lfcThreshold, na.rm=TRUE)

```

# Venn Diagrams

Generate Venn Diagrams to visualize gene expression between contrasts of interest

## Generate List of DE Genes

```{r}
# Gene list for contrast between weeks 12 and 9 (Week 9 is reference)

june_12v9 <- june_res_12v9[which(!is.na(june_res_12v9$padj) & june_res_12v9$padj < alpha & abs(june_res_12v9$log2FoldChange) > lfcThreshold),]@rownames
utah_12v9 <- utah_res_12v9[which(!is.na(utah_res_12v9$padj) & utah_res_12v9$padj < alpha & abs(utah_res_12v9$log2FoldChange) > lfcThreshold),]@rownames
straw_12v9 <- straw_res_12v9[which(!is.na(straw_res_12v9$padj) & straw_res_12v9$padj < alpha & abs(straw_res_12v9$log2FoldChange) > lfcThreshold),]@rownames

all_12v9 <- list(June = june_12v9, Utah = utah_12v9, Strawberry = straw_12v9)

# Gene list for contrast between weeks 14 and 9 (Week 9 is reference)

june_14v9 <- june_res_14v9[which(!is.na(june_res_14v9$padj) & june_res_14v9$padj < alpha & abs(june_res_14v9$log2FoldChange) > lfcThreshold),]@rownames
utah_14v9 <- utah_res_14v9[which(!is.na(utah_res_14v9$padj) & utah_res_14v9$padj < alpha & abs(utah_res_14v9$log2FoldChange) > lfcThreshold),]@rownames
straw_14v9 <- straw_res_14v9[which(!is.na(straw_res_14v9$padj) & straw_res_14v9$padj < alpha & abs(straw_res_14v9$log2FoldChange) > lfcThreshold),]@rownames

all_14v9 <- list(June = june_14v9, Utah = utah_14v9, Strawberry = straw_14v9)

# Gene list for contrast between weeks 14 and 12 (Week 12 is reference)

june_14v12 <- june_res_14v12[which(!is.na(june_res_14v12$padj) & june_res_14v12$padj < alpha & abs(june_res_14v12$log2FoldChange) > lfcThreshold),]@rownames
utah_14v12 <- utah_res_14v12[which(!is.na(utah_res_14v12$padj) & utah_res_14v12$padj < alpha & abs(utah_res_14v12$log2FoldChange) > lfcThreshold),]@rownames
straw_14v12 <- straw_res_14v12[which(!is.na(straw_res_14v12$padj) & straw_res_14v12$padj < alpha & abs(straw_res_14v12$log2FoldChange) > lfcThreshold),]@rownames

all_14v12 <- list(June = june_14v12, Utah = utah_14v12, Strawberry = straw_14v12)
```

## Generate Venn Diagrams

```{r}

library(ggplot2)
library(ggVennDiagram)
#library(ggpubr)
library(gridExtra)

getVenn <- function(list) {
  
  vennData <- Venn(list) %>% process_data()
  
  myPlot <- ggplot() +
    # 1. region count layer
    geom_sf(aes(fill = count), data = venn_region(vennData)) +
    # 2. set edge layer
    geom_sf(color = "grey", size = 3, data = venn_setedge(vennData),
            show.legend = FALSE) +
    # 3. set label layer
    geom_sf_text(aes(label = name), data = venn_setlabel(vennData)) +
    # 4. region label layer
    geom_sf_label(aes(label = count), data = venn_region(vennData)) +
    scale_fill_distiller(palette = "Blues", direction = 1) +
    theme_void() +
    theme(legend.position = "none")
  
  return(myPlot)
  
}

venn_12v9 <- getVenn(all_12v9)
venn_14v9 <- getVenn(all_14v9)
venn_14v12 <- getVenn(all_14v12)

myPlot <- grid.arrange(venn_12v9, venn_14v9, venn_14v12, ncol = 1)

ggsave("./figures/venn_diagram/venn_diagrams.pdf", myPlot)

```
